File: ruby/space-age/space_age.rb
class SpaceAge
  
  EARTH_YEAR_DURATION = 31557600

  PLANETS_WITH_YEAR_DURATION = { 
    mercury: EARTH_YEAR_DURATION * 0.2408467,
    venus:   EARTH_YEAR_DURATION * 0.61519726,
    earth:   EARTH_YEAR_DURATION,
    mars:    EARTH_YEAR_DURATION * 1.8808158,
    jupiter: EARTH_YEAR_DURATION * 11.862615,
    saturn:  EARTH_YEAR_DURATION * 29.447498,
    uranus:  EARTH_YEAR_DURATION * 84.016846,
    neptune: EARTH_YEAR_DURATION * 164.79132
   }.freeze


  PLANETS_WITH_YEAR_DURATION.each do |planet, year_duration|
    define_method "on_#{planet}" do 
      (seconds.to_f / year_duration).round(2)
    end
  end

  attr_reader :seconds
  def initialize(seconds)
    @seconds = seconds
  end

end

File: javascript/space-age/space-age.js
(function(){
  'use strict';

  function SpaceAge(seconds) {
    this.seconds = seconds;
    this.earthYearInSeconds = 31557600;
  }

  var planetsWithRelativeToEarthMultipliers = {
    "Earth": 1,
    "Mercury": 0.240846,
    "Venus": 0.61519726,
    "Mars": 1.8808158,
    "Jupiter": 11.862615,
    "Saturn": 29.447498,
    "Uranus": 84.016846,
    "Neptune": 164.79132
  }

  for (var planet in planetsWithRelativeToEarthMultipliers) {
    (function(planet) {
      SpaceAge.prototype["on" + planet] = function() {
        var multiplier = planetsWithRelativeToEarthMultipliers[planet];
        return this._calculate(multiplier);
      }
    })(planet);    
  }

  SpaceAge.prototype._calculate = function(multiplier) {
    var value = (this.seconds / this.earthYearInSeconds) / multiplier
    return this._round(value);
  };

  SpaceAge.prototype._round = function(value) {
    return Math.round(value * 100) / 100;
  };

  module.exports = SpaceAge;

})();